packages/core/src/publicCredential/PublicCredential.chain.ts

```diff
@@ -0,0 +1,230 @@
+  const api = ConfigService.get('api')
+
+  const { blockNumber, revoked } = publicCredentialEntry.unwrap()
+
+  const extrinsic = await retrievePublicCredentialCreationExtrinsicsFromBlock(
+    api,
+    credentialId,
+    blockNumber
+  )
+
+  if (extrinsic === null) {
+    throw new SDKErrors.PublicCredentialError(
+      `The block number as specified in the provided credential entry (${blockNumber}) does not have any extrinsic that includes a credential creation.`
+    )
+  }
+
+  if (!api.tx.did.submitDidCall.is(extrinsic)) {
+    throw new SDKErrors.PublicCredentialError(
+      'Extrinsic should be a did.submitDidCall extrinsic'
+    )
+  }
+
+  const [extrinsicCalls, extrinsicDidOrigin] = [
+    flattenCalls(api, extrinsic.args[0].call),
+    didFromChain(extrinsic.args[0].did),
+  ]
+
+  let credentialInput: IPublicCredential | undefined
+
+  // Iterate over the calls in the extrinsic to find the right one, and re-create the issued public credential.
+  // If more than a call are present, it always consider the last one as the valid one.
+  extrinsicCalls.forEach((call) => {
+    if (api.tx.publicCredentials.add.is(call)) {
+      const credentialCallArgument = call.args[0]
+      const reconstructedCredentialInput = credentialInputFromChain(
+        credentialCallArgument
+      )
+      const reconstructedId = computeId(
+        reconstructedCredentialInput,
+        extrinsicDidOrigin
+      )
+      if (reconstructedId === credentialId) {
+        credentialInput = {
+          ...reconstructedCredentialInput,
+          attester: extrinsicDidOrigin,
+          id: reconstructedId,
+          blockNumber,
+          revoked: revoked.toPrimitive(),
+        }
+      }
+    }
+  })
+  if (!credentialInput) {
+    throw new SDKErrors.PublicCredentialError(
+      'Block should always contain the full credential, eventually.'
+    )
+  }
+  return credentialInput
+}
+
+/**
+ * Decodes the public credential details returned by `api.call.publicCredential.getCredentials()`.
+ *
+ * This is the **only** secure way for users to retrieve and verify all the credentials issued to a given [[AssetDidUri]].
+ * Hence, calling `api.call.publicCredentials.getCredentials(asset_id)` and then passing the result to this function is the only way to trust that the credentials for a given AssetDID are valid.
+ *
+ * @param publicCredentialEntries The raw public credential details from blockchain.
+ * @returns An array of [[IPublicCredential]] as the result of combining the on-chain information and the information present in the tx history. If the result is an error, it maps it to the right error type.
+ */
+export async function credentialsFromChain(
+  publicCredentialEntries: Result<
+    Vec<ITuple<[Hash, PublicCredentialsCredentialsCredentialEntry]>>,
+    PublicCredentialError
+  >
+): Promise<IPublicCredential[]> {
+  if (publicCredentialEntries.isErr) {
+    throw new Error(publicCredentialEntries.asErr.toString())
+  }
```

Comment:
I believe the call of `toString()` is implicit here as well

Reply:
It is not, at least the compiler is complaining. `asErr` returns a `Codec` type, and it does not use `toString()` by default, AFAICT.

Reply:
Okay, I see